File: src/ui/bbcode/colors.ts
/**
 * `[color=…]` / `[size=…]` 的取值归一。
 *
 * NGA 编辑器给的是 CSS 颜色名与百分比字号（官方 `js_bbscode_core.js` 的
 * `ubbcode.fontColor` 24 色 + `ubbcode.fontSize` 五档），但正文里手打什么的都有。
 * 认不出来就返回 undefined —— 让那段文字按默认样式渲染，
 * 而不是把非法值塞进 style 里让 RN 在真机上炸掉。
 */

/**
 * 官方调色板的 24 色（`ubbcode.fontColor`），加上正文里常见但不在面板上的几个。
 * 全是 CSS 颜色关键字，RN 的颜色解析认得，所以只做白名单校验、不做换算。
 */
const NAMED_COLORS = new Set([
  // ubbcode.fontColor 原样 24 色
  'skyblue',
  'royalblue',
  'blue',
  'darkblue',
  'orange',
  'orangered',
  'crimson',
  'red',
  'firebrick',
  'darkred',
  'green',
  'limegreen',
  'seagreen',
  'teal',
  'deeppink',
  'tomato',
  'coral',
  'purple',
  'indigo',
  'burlywood',
  'sandybrown',
  'sienna',
  'chocolate',
  'silver',
  // 面板上没有、但老楼层里常见（white 是「防剧透」的写法，故意让人选中才看得见）
  'white',
  'black',
  'gray',
  'grey',
  'yellow',
  'pink',
  'brown',
  'navy',
  'olive',
  'maroon',
  'cyan',
  'magenta',
  'lightblue',
  'lightgreen',
  'darkgreen',
  'darkorange',
  'gold',
])

const HEX_COLOR = /^#(?:[0-9a-f]{3}|[0-9a-f]{6}|[0-9a-f]{8})$/

/** 认得出的颜色返回可直接进 style 的值，否则 undefined。 */
export function resolveBBColor(value: string): string | undefined {
  const color = value.trim().toLowerCase()
  if (HEX_COLOR.test(color)) return color
  return NAMED_COLORS.has(color) ? color : undefined
}

/** 字号缩放的上下限：再大撑破楼层卡片，再小认不出字。 */
const MIN_SCALE = 0.6
const MAX_SCALE = 2.5

/**
 * `[size=120%]` → 1.2 倍。官方只发百分比；裸数字按同样的百分比语义处理
 * （`[size=150]` 这种手打形式在老楼层里有）。认不出返回 undefined。
 */
export function resolveBBSizeScale(value: string): number | undefined {
  const match = /^(\d+(?:\.\d+)?)%?$/.exec(value.trim())
  if (match === null) return undefined
  const scale = Number(match[1]) / 100
  if (!Number.isFinite(scale) || scale <= 0) return undefined
  return Math.min(MAX_SCALE, Math.max(MIN_SCALE, scale))
}

File: app/src/main/kotlin/com/chasel/ng2n/ui/bbcode/Colors.kt
package com.chasel.ng2n.ui.bbcode

import androidx.compose.ui.graphics.Color

private val NAMED_COLORS: Map<String, Color> = mapOf(
  "skyblue" to Color(0xFF87CEEB),
  "royalblue" to Color(0xFF4169E1),
  "blue" to Color(0xFF0000FF),
  "darkblue" to Color(0xFF00008B),
  "orange" to Color(0xFFFFA500),
  "orangered" to Color(0xFFFF4500),
  "crimson" to Color(0xFFDC143C),
  "red" to Color(0xFFFF0000),
  "firebrick" to Color(0xFFB22222),
  "darkred" to Color(0xFF8B0000),
  "green" to Color(0xFF008000),
  "limegreen" to Color(0xFF32CD32),
  "seagreen" to Color(0xFF2E8B57),
  "teal" to Color(0xFF008080),
  "deeppink" to Color(0xFFFF1493),
  "tomato" to Color(0xFFFF6347),
  "coral" to Color(0xFFFF7F50),
  "purple" to Color(0xFF800080),
  "indigo" to Color(0xFF4B0082),
  "burlywood" to Color(0xFFDEB887),
  "sandybrown" to Color(0xFFF4A460),
  "sienna" to Color(0xFFA0522D),
  "chocolate" to Color(0xFFD2691E),
  "silver" to Color(0xFFC0C0C0),
  "white" to Color(0xFFFFFFFF),
  "black" to Color(0xFF000000),
  "gray" to Color(0xFF808080),
  "grey" to Color(0xFF808080),
  "yellow" to Color(0xFFFFFF00),
  "pink" to Color(0xFFFFC0CB),
  "brown" to Color(0xFFA52A2A),
  "navy" to Color(0xFF000080),
  "olive" to Color(0xFF808000),
  "maroon" to Color(0xFF800000),
  "cyan" to Color(0xFF00FFFF),
  "magenta" to Color(0xFFFF00FF),
  "lightblue" to Color(0xFFADD8E6),
  "lightgreen" to Color(0xFF90EE90),
  "darkgreen" to Color(0xFF006400),
  "darkorange" to Color(0xFFFF8C00),
  "gold" to Color(0xFFFFD700),
)

private val HEX_COLOR = Regex("""^#(?:[0-9a-f]{3}|[0-9a-f]{6}|[0-9a-f]{8})$""")

const val SPOILER_COLOR_NAME: String = "white"

fun resolveBBColor(value: String): Color? {
  val color = value.trim().lowercase()
  if (HEX_COLOR.matches(color)) return parseHexColor(color)
  return NAMED_COLORS[color]
}

fun isSpoilerColor(value: String): Boolean = value.trim().lowercase() == SPOILER_COLOR_NAME

private fun parseHexColor(hex: String): Color {
  val digits = hex.substring(1)
  return when (digits.length) {
    3 -> Color(
      red = digits[0].hexDigit() * 17,
      green = digits[1].hexDigit() * 17,
      blue = digits[2].hexDigit() * 17,
    )
    6 -> Color(
      red = digits.substring(0, 2).toInt(16),
      green = digits.substring(2, 4).toInt(16),
      blue = digits.substring(4, 6).toInt(16),
    )
    else -> Color(
      red = digits.substring(0, 2).toInt(16),
      green = digits.substring(2, 4).toInt(16),
      blue = digits.substring(4, 6).toInt(16),
      alpha = digits.substring(6, 8).toInt(16),
    )
  }
}

private fun Char.hexDigit(): Int = Character.digit(this, 16)

private const val MIN_SCALE = 0.6f
private const val MAX_SCALE = 2.5f

private val SIZE_VALUE = Regex("""^(\d+(?:\.\d+)?)%?$""")

fun resolveBBSizeScale(value: String): Float? {
  val match = SIZE_VALUE.matchEntire(value.trim()) ?: return null
  val scale = (match.groupValues[1].toFloatOrNull() ?: return null) / 100f
  if (!scale.isFinite() || scale <= 0f) return null
  return scale.coerceIn(MIN_SCALE, MAX_SCALE)
}
